gh-75593: Allow opening of path-like files with `wave.open`

mbyrnepr2 opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-140952: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->
https://github.com/python/cpython/issues/75593

<!-- gh-issue-number: gh-75593 -->
* Issue: gh-75593
<!-- /gh-issue-number -->

mbyrnepr2 committed:
wave.py - Allow opening of path-like files.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

mbyrnepr2 committed:
Delete Misc/NEWS.d/next/Library/2025-11-03-15-51-42.gh-issue-140952.EFVhKR.rst

mbyrnepr2 committed:
Update the `wave.py` documentation.

serhiy-storchaka commented on Lib/wave.py:
```diff
@@ -274,7 +275,7 @@ def initfp(self, file):
     def __init__(self, f):
         self._i_opened_the_file = None
-        if isinstance(f, str):
+        if isinstance(f, (str, os.PathLike)):
```
As was suggested in https://github.com/python/cpython/issues/75593#issuecomment-3494380746, it is easy to add support of `bytes`.

serhiy-storchaka commented on Lib/test/test_wave.py:
```diff
@@ -205,6 +207,18 @@ def test_open_in_write_raises(self):
+    def test_open_pathlike(self):
+        """It is possible to use `wave.read` and `wave.write` with a path-like file"""
+        with tempfile.NamedTemporaryFile(delete_on_close=False) as fp:
+            WAV_FILE = pathlib.Path(fp.name)
```
Use `test.support.os_helper.FakePath` instead of `pathlib.Path`. It should support arbitrary Path-like objects.

Also, test with a Path-like object wrapping a bytes path: `FakePath(os.fsencode(...))`.

serhiy-storchaka commented on Doc/library/wave.rst:
```diff
@@ -25,7 +25,7 @@ The :mod:`wave` module defines the following function and exception:
 .. function:: open(file, mode=None)
 
-   If *file* is a string, open the file by that name, otherwise treat it as a
+   If *file* is a string or a :term:`path-like object`, open the file by that name, otherwise treat it as a
```
Too long line.

Add also a `versionchanged` directive.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2025-11-04-12-16-13.gh-issue-75593.EFVhKR.rst:
```diff
@@ -0,0 +1 @@
+wave.py - Allow opening of path-like files.
```
"
```suggestion
Add support of :term:`path-like objects <path-like object>` in :func:`wave.open`.
```

mbyrnepr2 committed:
Update Misc/NEWS.d/next/Library/2025-11-04-12-16-13.gh-issue-75593.EFVhKR.rst

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

mbyrnepr2 committed:
Code review improvements:
- Use `FakePath` instead of `pathlib.Path`.
- Add an extra test case for the bytes path case.
- Format the doc and add `versionchanged`.

mbyrnepr2 committed:
Remove unused import.

serhiy-storchaka commented on Lib/test/test_wave.py:
```diff
@@ -206,5 +208,24 @@ def test_open_in_write_raises(self):
+        with tempfile.NamedTemporaryFile(delete_on_close=False) as fp:
+            cases = (
+                FakePath(fp.name),
+                FakePath(os.fsencode(fp.name)),
```
Test also for `os.fsencode(fp.name)`.

serhiy-storchaka commented on Lib/test/test_wave.py:
```diff
@@ -206,5 +208,24 @@ def test_open_in_write_raises(self):
             self.assertIsNone(cm.unraisable)
 
 
+class WaveOpen(unittest.TestCase):
```
Isn't there a place for this test in an existing class? You can also look in audiotests -- if this was added before removing aifc, the test should be added there.

mbyrnepr2 replied:
I had also considered adding the new tests to [MiscTestCase](https://github.com/python/cpython/blob/main/Lib/test/test_wave.py#L135) or [WaveLowLevelTest](https://github.com/python/cpython/blob/main/Lib/test/test_wave.py#L141) - I don't have a good reason to create a new class now that I think of it, except that perhaps they don't strictly belong in either of those places.

Just to clarify your point about adding the tests to audiotests.py - are you saying that the new tests should simply be moved there instead? Or do you mean that the _existing_ tests in test_wave.py should inherit the _new_ tests similar to the existing design?
My rationale for adding the tests to test_wave.py is I did feel like they belonged in test_wave.py and not audiotests.py since the classes in audiotests.py are inherited by the classes in test_wave.py to assert a range of things which felt redundant for the new tests to also perform.

serhiy-storchaka commented on Doc/library/wave.rst:
```diff
@@ -25,8 +25,8 @@ The :mod:`wave` module defines the following function and exception:
 
-   If *file* is a string, open the file by that name, otherwise treat it as a
-   file-like object.  *mode* can be:
+   If *file* is a string or a :term:`path-like object`, open the file by that name,
```
or a bytes object.

serhiy-storchaka commented on Doc/library/wave.rst:
```diff
@@ -52,6 +52,9 @@ The :mod:`wave` module defines the following function and exception:
       Added support for unseekable files.
 
+   .. versionchanged:: 3.15
+      Added support for :term:`path-like objects <path-like object>`.
```
and bytes objects.

mbyrnepr2 committed:
Use code review improvements:
- More accurate documentation.
- More complete test cases.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2025-11-04-12-16-13.gh-issue-75593.EFVhKR.rst:
```diff
@@ -0,0 +1 @@
+Add support of :term:`path-like objects <path-like object>` in :func:`wave.open`.
```
And bytes objects.

mbyrnepr2 replied:
Oh sorry for that again 😅. I’ll update later today.

serhiy-storchaka reviewed (approved):
LGTM. :+1:

mbyrnepr2 committed:
Update the news with the full information.
